### PR TITLE
Domain join v2

### PR DESCRIPTION
**Pull request by RogerBestMsft**
PR with updates from review.  Still need to determine folder names.

### CODE REVIEW

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    Import-Module Microsoft.PowerShell.SecretManagement
+    Import-Module Microsoft.PowerShell.SecretStore
+   
+    # Setup Log file
+    $LogFile = Join-Path $($env:Userprofile) "DJLog$(Get-Date -Format o | ForEach-Object { $_ -replace ":", "." }).txt"
+    New-Item -Path $logFile -ItemType File
+
+    # Check Windows 10 / 11 Operating system
+    if (!([System.Environment]::OSVersion.Version.Major -match "10" -or [System.Environment]::OSVersion.Version.Major -match "11")) {
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1]: Just an idea- maybe convert condition to `[System.Environment]::OSVersion.Version.Major -lt 10`

Comment (EMaher) [reply TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1]: If you want to add a check to verify Windows OS in case the customer is using PowerShell 7, you could use something like:  `$($PSVersionTable.PSVersion.Major -ge 7) -and $IsWindows`.  Or `$($env:OS).toLower() -match "windows"` for backwards compatible way.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        Write-LogFile "Requires Windows 10 or Windows 11."
+        exit
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1]: I noticed you are just exiting when a pre-condition is not met.  Do we want call `exit 1` to set a non-zero exit code since the script didn't accomplish the intended task?  (Not sure on best practice for this scenario.)

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    $MetaDataHeaders = @{"Metadata"="true"}
+    $vminfo = Invoke-RestMethod -Method GET -uri "http://169.254.169.254/metadata/instance?api-version=2018-10-01" -Headers $MetaDataHeaders
+
+    if ($vminfo.compute.vmScaleSetName -match "template"){
+        Write-Log "DomainJoin-StudentVM script was not run on a student vm."
+        exit
+    }
+
+    # Unlock vault
+    $passwordPath = Join-Path $($env:Userprofile) SecretStore.vault.credential
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:74]: Can you add a note that secret is saved in other scripts.  In case other people do what I did and went straight into the code and only skimmed the readme.md.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
+# Sample: Domain register or join Windows 10/11 lab VMs
+This sample provides scripts to help with two recommended options for registering/joining Azure Lab Services VMs to a domain:
+- **Students self-register/join**: Students can self-register (via [AAD register](https://learn.microsoft.com/azure/active-directory/devices/concept-azure-ad-register)) or self-join (via [AAD join](https://learn.microsoft.com/azure/active-directory/devices/concept-azure-ad-join-hybrid)) their lab VMs.  Even when students self-register/join their lab VMs, you need to rename the lab VMs to ensure they are unique before they are registerd/joined to AAD by students.  We provide [sample PowerShell scripts](#rename-student-vms-for-aad-register-and-aad-join) that rename student VMs to unique names.
+- **Domain join on behalf of students**: IT admins can domain join lab VMs on behalf of students using PowerShell to Hybrid AAD join.  We provide [sample PowerShell scripts](#hybrid-aad-join-vms-on-behalf-of-students) that perform the domain join.
+
+For more information on this sample, see the blog post **TODO - need to add link when published**.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:6]: I would argue that this sample should be self-contained and the blog post should point to this sample, not the sample point to the blog post.  Personally, I think blog posts are better at conveying point-in-time information rather than as reference docs.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+The steps in this section show how to set up the sample scripts that uniquely rename lab VMs.  For the full set of steps that show how students can self-register/join their lab VMs to AAD, see the blog post **TODO - need to add link when published**.
+
+### Prerequisites
+- **Lab creation**: The lab must be created with the **Create a template virtual machine** option enabled.  If you are planning students to AAD join their lab VMs, you should also enable the **Use same password for all virtual machines** option.
+- **Template VM setup**: You should install all other software/customizations *before* you set up the scripts to rename.  Also, we *don't* recommend exporting the template VM image with the scripts set up because they may cause ill-side effects (e.g., script inadvertently run on startup, errorneous values saved in the SecretStore, etc.)
+- **Windows versions**: The template and student VMs should use a Windows 10/11 image.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md]: What about Windows Server?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+The steps in this section show how to set up the sample scripts that uniquely rename lab VMs.  For the full set of steps that show how students can self-register/join their lab VMs to AAD, see the blog post **TODO - need to add link when published**.
+
+### Prerequisites
+- **Lab creation**: The lab must be created with the **Create a template virtual machine** option enabled.  If you are planning students to AAD join their lab VMs, you should also enable the **Use same password for all virtual machines** option.
+- **Template VM setup**: You should install all other software/customizations *before* you set up the scripts to rename.  Also, we *don't* recommend exporting the template VM image with the scripts set up because they may cause ill-side effects (e.g., script inadvertently run on startup, errorneous values saved in the SecretStore, etc.)
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:15]: Maybe rephrase in positive voice? 

If you plan to export the template VM image for future reuse, the template must be exported *before* these scripts are executed.  Scripts modify secret store and startup tasks for the machine and may cause unexpected side-affects.  Scripts can be saved locally on the template (but not executed) when template is exported for future use.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+- **Domain join on behalf of students**: IT admins can domain join lab VMs on behalf of students using PowerShell to Hybrid AAD join.  We provide [sample PowerShell scripts](#hybrid-aad-join-vms-on-behalf-of-students) that perform the domain join.
+
+For more information on this sample, see the blog post **TODO - need to add link when published**.
+
+## Rename VMs for student self-register/join to AAD
+Before a student can self-register/join their lab VM to AAD, the lab VM names should be both unique and meaningful to make it easier to manage them in AAD.  By default, Azure Labs doesn’t uniquely name lab VMs.  For example, if you enable the **Create a template virtual machine** option when you create the lab, the lab VMs will all be named like “lab000001”.  Or, if you leave this option disabled so that no template VM is created, the lab VMs will be named uniquely within a lab, but _not_ across labs.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:9]: So, what's the recommendation for labs w/o a template that users want to be self-registered?  Just save the script locally and ask student to run before registering the lab VM?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+- **Lab creation**: The lab must be created with the **Create a template virtual machine** option enabled.  If you are planning students to AAD join their lab VMs, you should also enable the **Use same password for all virtual machines** option.
+- **Template VM setup**: You should install all other software/customizations *before* you set up the scripts to rename.  Also, we *don't* recommend exporting the template VM image with the scripts set up because they may cause ill-side effects (e.g., script inadvertently run on startup, errorneous values saved in the SecretStore, etc.)
+- **Windows versions**: The template and student VMs should use a Windows 10/11 image.
+
+The steps in this sample typically need to be performed by an **IT admin** or **educator** who has:
+- Permission to [create labs](https://learn.microsoft.com/azure/lab-services/concept-lab-services-role-based-access-control#lab-creator-role) (at a minimum) in Azure Lab Services.
+- Proficiency with scripting, such as PowerShell.
+
+### Information you'll need
+- Lab name prefix. The VMs need unique names and the prefix will help identify the lab that the VMs belong to when viewing device entries in AD/AAD.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md]: Any limitations on prefix?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+- A SecretStore password.  This should be the password of a secondary local admin account that you add to the template VM, that is used to store values in the SecretStore.  You _shouldn't_ use the default local admin account because students that AAD join their lab VMs will need to use this account the first time they log into their lab VM.  Likewise, you may want students that use AAD register to also use the default local admin account.  For more info see the [Increasing security](#increasing-security) section.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:24]: If you really want to emphasize the second admin point, I would add it as a prerequisite.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+### Template VM details
+1.  Start the template VM and connect using the secondary local admin account.
+1.  Download the **Set-RenameValuesToSecretStoreTemplate.ps1** script.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:28]: I've found people appreciate when you add code to do this step for them
`Invoke-WebRequest 'https://raw.githubusercontent.com/Azure/LabServices/main/TemplateManagement/PowerShell/Set-RenameValuesToSecretStoreTemplate.ps1 -Outfile RenameValuesToSecretStoreTemplate.ps1`

Comment (EMaher) [reply TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:28]: Or you can add aka.ms links to make code less intimidating and add telemetry.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+- Creates a SecretStore with an encrypted password file.
+- Adds secrets to the SecretStore.
+- Downloads the **Rename-StudentVM.ps1** script that will be run on the student vm.
+
+On the template VM, the script will create a SecretStore that is specific to the current logged in user with the password entered.  The password is stored in an encrypted file specific to the user.  Additional secrets are added to the SecretStore to be accessed by the **Rename-StudentVM.ps1** script.
+
+### Publish and start all student VMs
+After you're done setting up the template VM:
+1.  Publish the lab.
+1.  Start all the student VMs which will trigger the **Rename-StudentVM.ps1** script to run on each student VM.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:45]: Stop the VMs after ___ minutes?  

How should user verify script worked?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+On the template VM, the script will create a SecretStore that is specific to the current logged in user with the password entered.  The password is stored in an encrypted file specific to the user.  Additional secrets are added to the SecretStore to be accessed by the **Rename-StudentVM.ps1** script.
+
+### Publish and start all student VMs
+After you're done setting up the template VM:
+1.  Publish the lab.
+1.  Start all the student VMs which will trigger the **Rename-StudentVM.ps1** script to run on each student VM.  
+
+Once the student VMs are renamed:
+- If the student is AAD joining their VM, they should log in the first time using the local admin account and follow the steps in the TODO: Link to blog post.  Once their lab VM is AAD joined, they can log in with their domain account by changing the RDP connection file accordingly.
+- If the student is AAD registering their VM, they can log in using either a local admin or non-admin accounnt and follow the steps in the TODO: Link to blog post.  Once their lab VM is AAD registered, they will continue to use the local admin or non-admin account to log in.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md]: typo - 'non-admin accounnt'

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+## Managing the template VM
+The template VM shouldn’t be joined to a domain because this image is used to create the student VMs and can be exported to create other labs.  You _can't_ reuse an image from a device that is already Hybrid AAD joined.  Instead, only the student VMs should be joined to a domain.  
+
+The **Join-Domain-StudentVM.ps1** script explicitly checks if it's being run on a student VM before it attempts to join to the domain.
+
+# Troubleshooting
+- The log file for the script is located in the logged in account's user folder.  The file will prefaced with "DJLog" followed by the date and time that the file was created.  All error messages will be logged there.
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:140]: It would be cool if you added a quick command to open or find the log on the VM.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+- If you are having issues with the script failing when run from a scheduled task, disable the task and try running the script manually from the student VM.  
+
+- Check that the student VM is properly joined to the domain.
+    - View the proper domain name in the AD/AAD UI.
+    - Open command window and use dsregcmd.  Additional troubleshooting tips using this tool. https://learn.microsoft.com/en-us/azure/active-directory/devices/troubleshoot-device-dsregcmd
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md]: Please remove 'en-us'

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+- If the password file or **Join-Domain-StudentVM.ps1** script has been removed from the student VM, you can reset the student VM or copy those files from the template VM.  If you also need to manually troubleshoot the script on the student VM, make sure you disable the scheduled startup task on the template VM (if applicable) before you reset the student VM; otherwise, the script will run automatically when you start the student VM.
+
+- If you've inadvertently rename or domain join the template VM, to reset the template back to it's original state you must _manually_ revert the changes.  To manually revert a domain join, you should remove the VM from the domain by changing from a domain to a workgroup named "WORKGROUP".  Once this has been completed, rename the machine to "Lab00000".
+
+# Increasing security
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:152]: I vote for secure by default and making a secondary admin account a prerequisite and adding this info into the known caveats.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+The instructions above use the VM's default local admin account.  If you want additional security, instead of using the default local admin account, you can add a secondary local adiministrator account (Security User) on the VM.  In this case, login to the template using the new Security User local account and run the **Set-DomainJoinValuesToSecretStore-TemplateVM.ps1**.  This will add an additional layer to help protect any secrets.  
+
+Students should log into their lab VM using their domain account; students _shouldn't_ be given access to the local accounts that have administrator permission.
+
+# Design Decisions
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:157]: FAQ?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    $pass = Import-CliXml $passwordPath
+    Unlock-SecretStore -Password $pass
+
+    # Check if vm renamed 
+    $computerName = (Get-WmiObject Win32_ComputerSystem).Name
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:79]: I'm curious why you us the WmiObject instead of environment variable like below.  (Honestly curious, nothing more.)

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+    $user = Get-LocalGroupMember -Member $aadGroup -Group $localGroup
+
+    if ((!$user) -and (((Get-WmiObject Win32_ComputerSystem).Domain) -ieq $Domain.toString())) {
+        Write-LogFile "Adding $aadGroup to $localGroup"
+        Add-LocalGroupMember -Group $localGroup -Member $aadGroup
+     }
+
+     Write-LogFile "Clean up section."
+     Remove-Item -Path $passwordPath -Force
+     Remove-Item -Path "C:\Users\Public\Documents\DomainJoin.ps1" -Force
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1]: You can get root of public path by using `$env:PUBLIC`.  It's different on server OSes.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 19]
```diff
@@ -0,0 +1,168 @@
+<#
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:1]: This filename doesn't follow PowerShell naming guidelines.  I think something like 'Join-LabVmToDomain' of 'Join-StudentVmToDomain' would fit the guidelines.  (If we care about following these guidelines.  I'm torn because 'DomainJoin-StudentVm' seems totally clear to me.)

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Rename-StudentVM.ps1 [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.  
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 
+.SYNOPSIS
+This script is part of the scripts chain for joining a student VM to an Active Directory domain. It renames the computer with a unique ID. Then it schedules the actual join script to run after reboot.
+.LINK https://docs.microsoft.com/en-us/azure/lab-services/classroom-labs/how-to-connect-peer-virtual-network
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Rename-StudentVM.ps1]: Please remove 'en-us'

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-DomainJoinValuesToSecretStore-TemplateVM.ps1 [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+#>
+
+
+[CmdletBinding()]
+param(
+    [Parameter(Mandatory,
+    ValueFromPipeline=$true, 
+    ValueFromPipelineByPropertyName=$true,
+    HelpMessage="Domain user that has join rights.")]
+    [ValidateNotNullOrEmpty()]
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-DomainJoinValuesToSecretStore-TemplateVM.ps1:19]: Love the help messages. :-)

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-RenameValuesToSecretStore-TemplateVM.ps1 [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+ 
+}
+
+Unlock-SecretStore -Password $pass
+
+Set-Secret -Name LabId -Secret $LabPrefix
+
+# Copy down files into the Public documents folder
+# TODO set the correct final location
+Invoke-WebRequest -Uri https://raw.githubusercontent.com/Azure/LabServices/domainjoinv2/TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Rename-StudentVM.ps1 -OutFile C:\Users\Public\Documents\Rename-StudentVM.ps1
```
Comment (EMaher) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-RenameValuesToSecretStore-TemplateVM.ps1]: Nice.

I mentioned this previously, but you can use $env:PUBLIC\Documents\Rename-StudentVM.ps1

**Review by EMaher — APPROVED** [review]

**Review by EMaher — APPROVED** [review]

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md (hunk 6 above)
Comment (rogerbest1999) [reply TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:15]: We don't want them to export the template.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 (hunk 17 above)
Comment (rogerbest1999) [reply TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:79]: No good reason, existing code I copied from.

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 23]
```diff
@@ -56,9 +56,9 @@ try {
     New-Item -Path $logFile -ItemType File
 
     # Check Windows 10 / 11 Operating system
-    if (!([System.Environment]::OSVersion.Version.Major -match "10" -or [System.Environment]::OSVersion.Version.Major -match "11")) {
+    if ([System.Environment]::OSVersion.Version.Major -lt "10") {
```
Comment (petehauge) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:59]: We should probably also confirm windows correct?  (since someone could try running the script on Linux via powershell core and if the OS version happens to line up here, like Ubuntu 22.04?)

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 24]
```diff
@@ -70,7 +70,7 @@ try {
         exit
```
Comment (petehauge) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:70]: We should use the full script name, "Join-Domain-StudentVM.ps1" and instead of stating what didn't happen, should list the requirement, "Join-Domain-StudentVM.ps1 script must be run on a student virtual machine in Lab Services"

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 [hunk 25]
```diff
@@ -138,7 +138,7 @@ try {
 
      Write-LogFile "Clean up section."
      Remove-Item -Path $passwordPath -Force
-     Remove-Item -Path "C:\Users\Public\Documents\DomainJoin.ps1" -Force
+     Remove-Item -Path "$env:PUBLIC\Documents\DomainJoin.ps1" -Force
```
Comment (petehauge) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1:141]: Where is this script, "DomainJoin.ps1"?  I searched the scripts but don't see that we're writing this?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-DomainJoinValuesToSecretStore-TemplateVM.ps1 [hunk 26]
```diff
@@ -134,5 +134,5 @@ Set-Secret -Name LabId -Secret $LabPrefix
 
 # Copy down files into the Public documents folder
 # TODO set the correct final location
-Invoke-WebRequest -Uri https://raw.githubusercontent.com/Azure/LabServices/domainjoinv2/TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 -OutFile C:\Users\Public\Documents\Join-Domain-StudentVM.ps1
+Invoke-WebRequest -Uri https://raw.githubusercontent.com/Azure/LabServices/domainjoinv2/TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Join-Domain-StudentVM.ps1 -OutFile "$env:PUBLIC\Documents\Join-Domain-StudentVM.ps1"
```
Comment (petehauge) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/Set-DomainJoinValuesToSecretStore-TemplateVM.ps1:137]: Do we need 2 scripts for setting values to the secret store?  Perhaps we could only set the values that were passed in?  I can see customers getting confused about which of the two scripts they need?

File: TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md [hunk 27]
```diff
@@ -13,14 +13,14 @@ The steps in this section show how to set up the sample scripts that uniquely re
 ### Prerequisites
 - **Lab creation**: The lab must be created with the **Create a template virtual machine** option enabled.  If you are planning students to AAD join their lab VMs, you should also enable the **Use same password for all virtual machines** option.
 - **Template VM setup**: You should install all other software/customizations *before* you set up the scripts to rename.  Also, we *don't* recommend exporting the template VM image with the scripts set up because they may cause ill-side effects (e.g., script inadvertently run on startup, errorneous values saved in the SecretStore, etc.)
-- **Windows versions**: The template and student VMs should use a Windows 10/11 image.
+- **Windows versions**: The template and student VMs should use a Windows 10/11 image.  This has not been tested on Windows server versions.
```
Comment (petehauge) [inline TemplateManagement/PowerShell/ActiveDirectoryJoinV2/readme.md:16]: There are really 2 scenarios here (as I'm reading through) which require different scripts.  Perhaps those each go in subdirectories for the scenario?  That way you could have a high-level readme and then just point to the subdirectories for each of the two scenarios.  It also might be interesting to use an aka.ms link to point at the sub directories to see how many people pick each path?

**Comment by EMaher** [comment]
Closing PR.  See https://techcommunity.microsoft.com/t5/azure-lab-services-blog/use-labs-without-registering-joining-to-ad-aad/ba-p/3880261 for details.
